server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/remoteDelivery/DeliveryRunnable.java

```diff
@@ -196,264 +166,43 @@ public void run() {
-            }
-            MimeMessage message = mail.getMessage();
-
-            // Create an array of the recipients as InternetAddress objects
-            Collection<MailAddress> recipients = mail.getRecipients();
-            InternetAddress addr[] = new InternetAddress[recipients.size()];
-            int j = 0;
-            for (Iterator<MailAddress> i = recipients.iterator(); i.hasNext(); j++) {
-                MailAddress rcpt = i.next();
-                addr[j] = rcpt.toInternetAddress();
-            }
-
-            if (addr.length <= 0) {
-                logger.info("No recipients specified... not sure how this could have happened.");
-                return true;
-            }
-
-            // Figure out which servers to try to send to. This collection
-            // will hold all the possible target servers
-            Iterator<HostAddress> targetServers;
-            if (configuration.getGatewayServer().isEmpty()) {
-                MailAddress rcpt = recipients.iterator().next();
-                String host = rcpt.getDomain();
-
-                // Lookup the possible targets
-                try {
-                    targetServers = new MXHostAddressIterator(dnsServer.findMXRecords(host).iterator(), dnsServer, false, logger);
-                } catch (TemporaryResolutionException e) {
-                    logger.info("Temporary problem looking up mail server for host: " + host);
-                    String exceptionBuffer = "Temporary problem looking up mail server for host: " + host + ".  I cannot determine where to send this message.";
-
-                    // temporary problems
-                    return failMessage(mail, new MessagingException(exceptionBuffer), false);
-                }
-                if (!targetServers.hasNext()) {
-                    logger.info("No mail server found for: " + host);
-                    String exceptionBuffer = "There are no DNS entries for the hostname " + host + ".  I cannot determine where to send this message.";
-
-                    int retry = 0;
-                    try {
-                        retry = Integer.parseInt(mail.getErrorMessage());
-                    } catch (NumberFormatException e) {
-                        // Unable to parse retryCount
-                    }
-                    if (retry == 0 || retry > configuration.getDnsProblemRetry()) {
-                        // The domain has no dns entry.. Return a permanent
-                        // error
-                        return failMessage(mail, new MessagingException(exceptionBuffer), true);
-                    } else {
-                        return failMessage(mail, new MessagingException(exceptionBuffer), false);
-                    }
-                }
-            } else {
-                targetServers = getGatewaySMTPHostAddresses(configuration.getGatewayServer());
+            Boolean host = tryDeliver(mail, session);
```

Comment:
It do not persist in the later code. It's the simplest adaptation I found for extracting (which was pretty hard, sorry, I could not keep it exactly as is because of flow (continue, return, throws, reallocation)). 

I did the simplest refactoring steps to avoid side effects and allow extracts.